File: template/ios/IceCreamTemplate/Interface/RNSyncEngine.m
//
//  RNSyncEngine.m
//  IceCreamTemplate
//
//  Created by Leonid Zolotarev on 1/8/22.
//

#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE (RNSyncEngine, NSObject)

RCT_EXTERN_METHOD(startEngine)
RCT_EXTERN_METHOD(pull)
RCT_EXTERN_METHOD(pushAll)

@end

File: template/ios/IceCreamTemplate/Interface/RNSyncEngine.swift
//
//  RNSyncEngine.swift
//  IceCreamTemplate
//
//  Created by Leonid Zolotarev on 1/8/22.
//

import Foundation
import RealmSwift
import IceCream

@objc(RNSyncEngine)
class RNSyncEngine : NSObject {

  private var _syncEngine: SyncEngine? = nil

  @objc
  static func requiresMainQueueSetup() -> Bool {
      return false
  }

  @objc
  func startEngine() -> Void {
    DispatchQueue.main.async {
      self._syncEngine = SyncEngine(objects: [
        SyncObject(type: Task.self),
      ])
    }
  }

  @objc
  func pull() -> Void {
    DispatchQueue.main.async {
      self._syncEngine?.pull()
    }
  }
  
  @objc
  func pushAll() -> Void {
    DispatchQueue.main.async {
      self._syncEngine?.pushAll()
    }
  }
}
